### PR TITLE
Mam wots sign

### PR DESCRIPTION
**Pull request by tsvisabo**
- wots_sign - flex_trit_t
- wots_calc_ps - complete flex_trit_t transition to caller functions

### CODE REVIEW

File: mam/v2/mss.c [hunk 1]
```diff
@@ -65,7 +65,11 @@ static void mss_mt_gen_leaf(mss_t *mss, mss_mt_index_t index, trits_t h) {
   trits_put18(Ni, index);
   wots_gen_sk3(mss->wots, mss->prng, mss->nonce1, mss->nonce2, Ni);
   // calc pk & push hash
-  wots_calc_pk(mss->wots, h);
+  TRIT_ARRAY_MAKE_FROM_RAW(pk_trits_array, MAM2_WOTS_PK_SIZE, pk.p + pk.d);
```
Comment (thibault-martinez) [inline mam/v2/mss.c]: Can we use `TRIT_ARRAY_DECLARE` instead ?

File: mam/v2/mss.c [hunk 2]
```diff
@@ -393,7 +397,13 @@ void mss_sign(mss_t *mss, trits_t H, trits_t sig) {
     wots_gen_sk3(mss->wots, mss->prng, mss->nonce1, mss->nonce2, Ni);
   }
 
-  wots_sign(mss->wots, H, trits_take(sig, MAM2_WOTS_SIG_SIZE));
+  flex_trits_to_trits(sig.p + sig.d, MAM2_WOTS_SK_SIZE, mss->wots->sk,
+                      MAM2_WOTS_SK_SIZE, MAM2_WOTS_SK_SIZE);
+  TRIT_ARRAY_MAKE_FROM_RAW(hash_array, MAM2_WOTS_HASH_SIZE, hash.p + hash.d);
+  TRIT_ARRAY_MAKE_FROM_RAW(sk_sig_array, MAM2_WOTS_SIG_SIZE, sig.p + sig.d);
```
Comment (thibault-martinez) [inline mam/v2/mss.c:417]: Can we use TRIT_ARRAY_DECLARE instead ?

File: mam/v2/tests/test.c [hunk 3]
```diff
@@ -136,11 +136,20 @@ void test_wots_gen_sign(size_t Kn, char *K, size_t Nn, char *N, size_t pkn,
   prng_init(p, s, key);
   trytes_to_trits(N, tN.p, MIN(strlen(N), tN.n / RADIX));
   wots_gen_sk(w, p, tN);
-  wots_calc_pk(w, tpk);
+  TRIT_ARRAY_MAKE_FROM_RAW(pk_trits_array, MAM2_WOTS_PK_SIZE, tpk.p + tpk.d);
```
Comment (thibault-martinez) [inline mam/v2/tests/test.c:124]: Can we use TRIT_ARRAY_DECLARE instead ?

File: mam/v2/tests/test.c [hunk 4]
```diff
@@ -136,11 +136,20 @@ void test_wots_gen_sign(size_t Kn, char *K, size_t Nn, char *N, size_t pkn,
...
+  flex_trits_to_trits(tpk.p + tpk.d, MAM2_WOTS_PK_SIZE, pk_trits_array.trits,
+                      MAM2_WOTS_PK_SIZE, MAM2_WOTS_PK_SIZE);
   trits_to_trytes(tpk.p, pk, MIN(strlen(pk), tpk.n / RADIX));
 
   trytes_to_trits(H, tH.p, MIN(strlen(H), tH.n));
-  wots_sign(w, tH, tsig);
+  flex_trits_to_trits(tsig.p + tsig.d, MAM2_WOTS_SK_SIZE, w->sk,
+                      MAM2_WOTS_SK_SIZE, MAM2_WOTS_SK_SIZE);
+  TRIT_ARRAY_MAKE_FROM_RAW(hash_array, MAM2_WOTS_HASH_SIZE, tH.p + tH.d);
+  TRIT_ARRAY_MAKE_FROM_RAW(sk_sig_array, MAM2_WOTS_SIG_SIZE, tsig.p + tsig.d);
```
Comment (thibault-martinez) [inline mam/v2/tests/test.c]: Can we use TRIT_ARRAY_DECLARE instead ?

File: mam/v2/tests/test_wots.c [hunk 5]
```diff
@@ -42,8 +42,19 @@ bool_t wots_test_do(wots_t *w, prng_t *p) {
 
   prng_gen(p, 7, N, H);
   wots_gen_sk(w, p, N);
-  wots_calc_pk(w, pk);
-  wots_sign(w, H, sig);
+  TRIT_ARRAY_MAKE_FROM_RAW(pk_trits_array, MAM2_WOTS_PK_SIZE, pk.p + pk.d);
```
Comment (thibault-martinez) [inline mam/v2/tests/test_wots.c:46]: Can we use TRIT_ARRAY_DECLARE instead ?

File: mam/v2/tests/test_wots.c [hunk 6]
```diff
@@ -42,8 +42,19 @@ bool_t wots_test_do(wots_t *w, prng_t *p) {
...
+  wots_calc_pk(w, &pk_trits_array);
+  flex_trits_to_trits(pk.p + pk.d, MAM2_WOTS_PK_SIZE, pk_trits_array.trits,
+                      MAM2_WOTS_PK_SIZE, MAM2_WOTS_PK_SIZE);
+
+  flex_trits_to_trits(sig.p + sig.d, MAM2_WOTS_SK_SIZE, w->sk,
+                      MAM2_WOTS_SK_SIZE, MAM2_WOTS_SK_SIZE);
+  TRIT_ARRAY_MAKE_FROM_RAW(hash_array, MAM2_WOTS_HASH_SIZE, H.p + H.d);
+  TRIT_ARRAY_MAKE_FROM_RAW(sk_sig_array, MAM2_WOTS_SIG_SIZE, sig.p + sig.d);
```
Comment (thibault-martinez) [inline mam/v2/tests/test_wots.c:53]: Can we use TRIT_ARRAY_DECLARE instead ?

File: mam/v2/wots.c [hunk 7]
```diff
@@ -36,39 +38,51 @@ static void wots_calc_pks(isponge *const sponge, trit_array_p sk_pks,
...
 
-  for (i = 0; i < MAM2_WOTS_SK_PART_COUNT - 3; ++i) {
-    sig_part = trits_take(sk_sig, MAM2_WOTS_SK_PART_SIZE);
-    sk_sig = trits_drop(sk_sig, MAM2_WOTS_SK_PART_SIZE);
+  MAM2_ASSERT(sk_sig->num_trits == MAM2_WOTS_SK_SIZE);
+  MAM2_ASSERT(hash->num_trits == MAM2_WOTS_HASH_SIZE);
 
-    h = trits_get3(trits_drop(hash, i * 3));
+  TRIT_ARRAY_DECLARE(sk_sig_part, MAM2_WOTS_SK_PART_SIZE);
+  sk_sig_part.dynamic = 0;
```
Comment (thibault-martinez) [inline mam/v2/wots.c]: `TRIT_ARRAY_DECLARE ` is already doing that

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/v2/mss.c (hunk 2 above)
Comment (tsvisabo) [reply mam/v2/mss.c:417]: sig is initialized with `sk` but for hash, yes

File: mam/v2/tests/test_wots.c (hunk 6 above)
Comment (tsvisabo) [reply mam/v2/tests/test_wots.c:53]: for hash, thanks

File: mam/v2/mss.c (hunk 2 above)
Comment (thibault-martinez) [reply mam/v2/mss.c:417]: Oops, missed that

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**